File: benchmarks/python/1_fasta_stats.py
"""Benchmark 1: FASTA Statistics"""
from Bio import SeqIO
from statistics import mean, median

records = list(SeqIO.parse("data/sequences.fa", "fasta"))
lengths = [len(r.seq) for r in records]
total_bp = sum(lengths)
gc_values = [(r.seq.count("G") + r.seq.count("C")) / len(r.seq) for r in records]

# N50
sorted_desc = sorted(lengths, reverse=True)
cumsum = 0
n50 = 0
for l in sorted_desc:
    cumsum += l
    if cumsum >= total_bp / 2:
        n50 = l
        break

print(f"Sequences: {len(records)}")
print(f"Total bp: {total_bp}")
print(f"Mean length: {mean(lengths):.1f}")
print(f"Median length: {median(lengths):.1f}")
print(f"Min length: {min(lengths)}")
print(f"Max length: {max(lengths)}")
print(f"Mean GC: {mean(gc_values):.4f}")
print(f"N50: {n50}")

File: benchmarks/language/sequence_io/r/1_fasta_stats.R
# Benchmark 1: FASTA Statistics
library(Biostrings)

seqs <- readDNAStringSet("data/sequences.fa")
lengths <- width(seqs)
total_bp <- sum(lengths)
gc_values <- letterFrequency(seqs, "GC", as.prob = TRUE)[, 1]

# N50
sorted_desc <- sort(lengths, decreasing = TRUE)
cumsum_vals <- cumsum(sorted_desc)
n50 <- sorted_desc[which(cumsum_vals >= total_bp / 2)[1]]

cat(sprintf("Sequences: %d\n", length(seqs)))
cat(sprintf("Total bp: %d\n", total_bp))
cat(sprintf("Mean length: %.1f\n", mean(lengths)))
cat(sprintf("Median length: %.1f\n", median(lengths)))
cat(sprintf("Min length: %d\n", min(lengths)))
cat(sprintf("Max length: %d\n", max(lengths)))
cat(sprintf("Mean GC: %.4f\n", mean(gc_values)))
cat(sprintf("N50: %d\n", n50))
